### PR TITLE
add MORPHEUS

### PR DESCRIPTION
**Pull request by jxmorris12**
adds MORPHEUS from the paper "It's Morphin' Time! Combating Linguistic Discrimination with Inflectional Perturbations" 

paper link: https://www.aclweb.org/anthology/2020.acl-main.263/
Github link: https://github.com/salesforce/morpheus

main components:
- new transformation: `WordSwapInflections`
- new goal function: `MinimizeBleu`
- new attack recipe: `MorpheusTan2020`

### CODE REVIEW

File: textattack/attack_recipes/faster_genetic_algorithm_jia_2019.py [hunk 1]
```diff
@@ -19,68 +19,69 @@ def FasterGeneticAlgorithmJia2019(model):
...
-    relative to the current altered version of the input. This results in a
-    hard-to-define attack surface, as the same word can be replaced many
-    times, leading to semantic drift. We instead pre-compute the allowed
-    substitutions S(x, i) at index i based on the original x. We define
-    S(x, i) as the set of x_i ∈ N(x_i) such that where probabilities are
-    assigned by a pre-trained language model, and the window radius W and
-    threshold δ are hyperparameters. We use W = 6 and δ = 5.
     """
-    # # @TODO update all this stuff
+    #
```
Comment (uvafan) [inline textattack/attack_recipes/faster_genetic_algorithm_jia_2019.py:23]: Did you mean to change this?

File: textattack/attack_recipes/morpheus_tan_2020.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    RepeatModification,
+    StopwordModification,
+)
+from textattack.goal_functions import MinimizeBleu
+from textattack.search_methods import GreedySearch
+from textattack.shared.attack import Attack
+from textattack.transformations import WordSwapInflections
+
+
+def MorpheusTan2020(model, goal_function="non_overlapping"):
```
Comment (uvafan) [inline textattack/attack_recipes/morpheus_tan_2020.py:11]: unused `goal_function` argument

File: textattack/attack_recipes/morpheus_tan_2020.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    # Swap words with their inflections
+    transformation = WordSwapInflections()
+
+    #
+    # Don't modify the same word twice or stopwords
+    #
+    constraints = [RepeatModification(), StopwordModification()]
+
+    #
+    # Greedily swap words (see psueucode, Algorithm 1 of the paper).
```
Comment (uvafan) [inline textattack/attack_recipes/morpheus_tan_2020.py:34]: misspelled pseudocode

File: textattack/commands/train_model/run_training.py [hunk 4]
```diff
@@ -25,8 +25,7 @@
 
 
 def _save_args(args, save_path):
-    """
-    Dump args dictionary to a json
+    """Dump args dictionary to a json.
```
Comment (uvafan) [inline textattack/commands/train_model/run_training.py:28]: Did you mean to change these? I guess this is a `make format` thing? But then why was it merged without formatting?

File: textattack/goal_functions/text/minimize_bleu.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
+import functools
+
+import nltk
+
+import textattack
+
+from .text_to_text_goal_function import TextToTextGoalFunction
+
+
+class MinimizeBleu(TextToTextGoalFunction):
```
Comment (uvafan) [inline textattack/goal_functions/text/minimize_bleu.py:10]: Capitalize BLEU? And maybe include link to BLEU paper?

File: textattack/goal_functions/text/minimize_bleu.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+        self.target_bleu = target_bleu
+        super().__init__(*args, **kwargs)
+
+    def _is_goal_complete(self, model_output, _):
+        bleu_score = 1.0 - self._get_score(model_output, _)
+        return bleu_score <= (self.target_bleu + MinimizeBleu.EPS)
+
+    def _get_score(self, model_output, _):
+        model_output_at = textattack.shared.AttackedText(model_output)
+        ground_truth_at = textattack.shared.AttackedText(self.ground_truth_output)
```
Comment (uvafan) [inline textattack/goal_functions/text/minimize_bleu.py:31]: Maybe override `init_attack_example` to save `self.ground_truth_at` so you don't have to cast it over and over again?

File: textattack/transformations/word_swap_inflections.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+        replacement_inflections_dict = lemminflect.getAllLemmas(word)
+        # `lemminflect.getAllLemmas` returns a dict mapping part-of-speech
+        # to available inflections. First, map part-of-speech from flair
+        # POS tag to lemminflect.
+        lemminflect_pos = self._flair_to_lemminflect_pos_map[word_part_of_speech]
+        return replacement_inflections_dict.get(lemminflect_pos, None)
+
+    def _get_transformations(self, current_text, indices_to_modify):
+        words = current_text.words
+        sentence = Sentence(" ".join(words))
```
Comment (uvafan) [inline textattack/transformations/word_swap_inflections.py:38]: why can't you pass in `current_text.text`?

File: textattack/transformations/word_swap_inflections.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+            )
+            transformed_texts_idx = []
+            for r in replacement_words:
+                transformed_texts_idx.append(current_text.replace_word_at_index(i, r))
+            transformed_texts.extend(transformed_texts_idx)
+
+        return transformed_texts
+
+
+def recover_word_case(word, reference_word):
```
Comment (uvafan) [inline textattack/transformations/word_swap_inflections.py:61]: is this method unused?

File: textattack/transformations/word_swap_inflections.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+        ), "Part-of-speech tagger returned incorrect number of tags"
+        transformed_texts = []
+
+        for i in indices_to_modify:
+            word_to_replace = words[i]
+            word_to_replace_pos = pos_list[i][:2]  # get the root POS
+            replacement_words = (
+                self._get_replacement_words(word_to_replace, word_to_replace_pos) or []
+            )
+            transformed_texts_idx = []
```
Comment (uvafan) [inline textattack/transformations/word_swap_inflections.py:53]: what is the purpose of `transformed_text_idx`?

File: textattack/goal_functions/text/minimize_bleu.py (hunk 6 above)
Comment (jxmorris12) [reply textattack/goal_functions/text/minimize_bleu.py:31]: eh, I don't think it's worth it :)

File: textattack/transformations/word_swap_inflections.py (hunk 7 above)
Comment (jxmorris12) [reply textattack/transformations/word_swap_inflections.py:38]: you can, this is better

File: docs/attacks/attack_recipes.rst [hunk 10]
```diff
@@ -4,6 +4,11 @@ Attack Recipes
 We provide a number of pre-built attack recipes. To run an attack recipe, run::
 
     textattack attack --recipe [recipe_name]
+
+
+Attacks on classification
```
Comment (uvafan) [inline docs/attacks/attack_recipes.rst:9]: nit: says sequence-to-sequence models below but no "models" here

**Pull request merged**